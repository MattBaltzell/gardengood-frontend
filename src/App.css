.App {
  text-align: center;
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
  padding: 2rem;
  background-color: #282c34;
  color: rgba(255, 255, 255, 0.6);
}

.container {
  display: grid;
  grid-template-columns: repeat(6, 1fr);
  gap: 2rem;
  width: 100%;
  max-width: 1200px;
}

.Plant {
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: left;
  justify-content: flex-start;
  width: 100%;
  text-align: center;
  /* background-color: #3d4553; */
  gap: 1rem;
  border-radius: 3px;
  /* box-shadow: 0 4px 8px rgba(0, 0, 0, 0.2); */

  border-radius: 50%;
}

.Plant h3 {
  font-size: 1rem;
  transition: all 0.2s ease-in-out;
}

.Plant-img {
  object-fit: cover;
  width: 100%;
  border-radius: 50%;
  box-shadow: 0 4px 8px rgba(0, 0, 0, 0.3);
  position: relative;
  transform: translateY(0);
  transition: all 0.2s ease-in-out;
}

.Plant-img:hover {
  filter: brightness(1.1);
  transform: translateY(-0.5rem) scale(1.05);
  box-shadow: 0 8px 16px rgba(0, 0, 0, 0.4);
}
.Plant:hover h3 {
  color: #fff;
  transform: translateY(-0.3rem) scale(1.04);
}
