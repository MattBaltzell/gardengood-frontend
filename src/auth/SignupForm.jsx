import React from "react";
import useFormFields from "../hooks/useFormFields";
import { useNavigate, Link } from "react-router-dom";
import "../components/Form/Form.css";

const SignupForm = ({ signup }) => {
  const INITIAL_STATE = {
    username: "",
    password: "",
    firstName: "",
    lastName: "",
    email: "",
    zipCode: "",
  };
  const { formData, setFormData, handleChange } = useFormFields(INITIAL_STATE);

  const navigate = useNavigate();

  const handleSubmit = (e) => {
    e.preventDefault();
    signup(formData);
    setFormData(INITIAL_STATE);
    navigate("/");
  };

  return (
    <main className="Form">
      <div className="Home__bg"></div>

      <h1>Sign Up</h1>
      <form onSubmit={handleSubmit}>
        <label htmlFor="username">Username</label>
        <input
          id="username"
          name="username"
          type="text"
          onChange={handleChange}
          value={formData.username}
        />
        <label htmlFor="password">Password</label>
        <input
          id="password"
          name="password"
          type="password"
          onChange={handleChange}
          value={formData.password}
        />
        <label htmlFor="firstName">First name</label>
        <input
          id="firstName"
          name="firstName"
          type="text"
          onChange={handleChange}
          value={formData.firstName}
        />
        <label htmlFor="lastName">Last name</label>
        <input
          id="lastName"
          name="lastName"
          type="text"
          onChange={handleChange}
          value={formData.lastName}
        />
        <label htmlFor="email">Email</label>
        <input
          id="email"
          name="email"
          type="email"
          onChange={handleChange}
          value={formData.email}
        />
        <label htmlFor="zipCode">Zip Code</label>
        <input
          id="zipCode"
          name="zipCode"
          type="text"
          pattern="[0-9]{5}"
          onChange={handleChange}
          value={formData.zipCode}
        />
        <button>Create Account</button>
      </form>
      <p>
        Already have an account? <Link to="/login">Log in</Link>
      </p>
    </main>
  );
};

export default SignupForm;
