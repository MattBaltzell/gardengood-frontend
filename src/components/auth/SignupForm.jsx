import React, { useContext, useState } from "react";
import { Link, useNavigate, Navigate } from "react-router-dom";
import "../form/Form.css";
import { Formik, Form } from "formik";
import * as Yup from "yup";
import { MyTextInput, MyPasswordInput } from "../form/Forms";
import UserContext from "./UserContext";
import { v4 as uuid } from "uuid";

const SignupForm = ({ signup, toast }) => {
  // Pass the useFormik() hook initial form values, a validate function that will be called when
  // form values change or fields are blurred, and a submit function that will
  // be called when the form is submitted
  const navigate = useNavigate();
  const currUser = useContext(UserContext);
  const [submissionErrors, setSubmissionErrors] = useState([]);

  const INITIAL_STATE = {
    username: "",
    password: "",
    firstName: "",
    lastName: "",
    email: "",
    zipCode: "",
  };

  return currUser ? (
    <Navigate to="/" />
  ) : (
    <main className="Form">
      <div className="Home__bg"></div>

      <h1>Sign Up</h1>

      <Formik
        initialValues={{ ...INITIAL_STATE }}
        validationSchema={Yup.object({
          username: Yup.string()
            .max(25, "Must be 25 characters or less")
            .required("Required"),
          password: Yup.string()
            .min(8, "Must be 8 characters or more")
            .required("Required"),
          firstName: Yup.string()
            .max(25, "Must be 25 characters or less")
            .required("Required"),
          lastName: Yup.string()
            .max(30, "Must be 30 characters or less")
            .required("Required"),
          email: Yup.string()
            .email("Invalid email address")
            .required("Required"),
          zipCode: Yup.string()
            .length(5, "Must be exactly 5 characters")
            .required("Required"),
        })}
        onSubmit={(values) => {
          setSubmissionErrors([]);
          setTimeout(async () => {
            const res = await signup(values);
            if (res.message) {
              setSubmissionErrors((errors) => [...errors, res.message]);
            }
          }, 400);
        }}
      >
        <Form>
          <MyTextInput
            label="Username"
            name="username"
            type="text"
            autoComplete="username"
          />
          <MyTextInput
            label="Email"
            name="email"
            type="text"
            autoComplete="email"
          />
          <MyPasswordInput label="Password" name="password" type="password" />
          <MyTextInput label="First Name" name="firstName" type="text" />
          <MyTextInput label="Last Name" name="lastName" type="text" />
          <MyTextInput
            label="Zip Code"
            name="zipCode"
            type="text"
            pattern="[0-9]{5}"
          />
          {submissionErrors
            ? submissionErrors.map((e) => (
                <div key={uuid()} className="error center">
                  {e}
                </div>
              ))
            : null}
          <button type="submit">Create Account</button>
        </Form>
      </Formik>
      <p>
        Already have an account? <Link to="/login">Log in</Link>
      </p>
    </main>
  );
};

export default SignupForm;
