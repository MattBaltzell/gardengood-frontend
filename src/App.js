import { useState, useEffect } from "react";
import "./App.css";
import axios from "axios";

import PlantGrid from "./components/Plant/PlantGrid";
function App() {
  const [plants, setPlants] = useState([]);

  useEffect(() => {
    async function getPlants() {
      try {
        const res = await axios({
          url: "http://localhost:3001/plants",
          method: "GET",
          headers: {
            Authorization: `Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6InRlc3RhZG1pbiIsImlzQWRtaW4iOmZhbHNlLCJpYXQiOjE2Nzc2NDA3Mzl9.HoNDrjYXseNCN6CIKkCB9FqT6ecJumAVNE6OeTg1WLk`,
          },
        });
        const plants = res.data.plants;

        setPlants(plants);
      } catch (error) {
        console.error(error);
      }
    }
    getPlants();
  }, []);
  return (
    <div className="App">
      <div className="container">
        <PlantGrid plants={plants} />
      </div>
    </div>
  );
}

export default App;
