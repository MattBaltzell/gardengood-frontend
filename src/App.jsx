import React, { useState, useEffect } from "react";
import "./App.css";
import Router from "./components/Navigation-Routes/Router";
import Navbar from "./components/Navigation-Routes/Navbar";
import GardenGoodApi from "./api/api";
import UserContext from "./auth/UserContext";
import jwt_decode from "jwt-decode";

function App() {
  const [currUser, setCurrUser] = useState(null);
  const [token, setToken] = useState(localStorage.getItem("gardengood-token"));
  const [userWasUpdated, setUserWasUpdated] = useState(false);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    const getCurrUser = async () => {
      if (token) {
        try {
          GardenGoodApi.token = token;
          let { username } = jwt_decode(token);
          const user = await GardenGoodApi.getCurrentUser(username);
          setCurrUser(user);
          localStorage.setItem("gardengood-token", token);
        } catch (error) {
          console.error(error);
        }
      }
    };
    getCurrUser();

    setUserWasUpdated(false);
  }, [token, userWasUpdated]);

  const handleIsLoading = (bool) => {
    setIsLoading(bool);
  };

  const login = async ({ username, password }) => {
    try {
      const token = await GardenGoodApi.login({ username, password });
      setToken(token);
    } catch (error) {
      return { message: error };
    }
  };

  const signup = async ({
    username,
    password,
    firstName,
    lastName,
    email,
    zipCode,
  }) => {
    try {
      const token = await GardenGoodApi.signup({
        username,
        password,
        firstName,
        lastName,
        email,
        zipCode,
      });
      setToken(token);
    } catch (error) {
      return { message: error };
    }
  };

  const logout = async () => {
    try {
      setToken(null);
      setCurrUser(null);
      GardenGoodApi.token = null;
      localStorage.setItem("gardengood-token", "");
    } catch (error) {
      return { message: error };
    }
  };

  return (
    <div className="App">
      <UserContext.Provider value={currUser}>
        <Navbar logout={logout} />
        <Router
          login={login}
          signup={signup}
          logout={logout}
          handleIsLoading={handleIsLoading}
          isLoading={isLoading}
        />
      </UserContext.Provider>
    </div>
  );
}

export default App;
