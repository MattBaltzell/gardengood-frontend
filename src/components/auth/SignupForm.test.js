import React from "react";
import {
  fireEvent,
  render,
  screen,
  waitFor,
  act,
} from "@testing-library/react";
import { MemoryRouter } from "react-router";
import SignupForm from "./SignupForm";
import App from "../../App";
import userEvent from "@testing-library/user-event";

describe("SignupForm component tests", function () {
  it("renders without crashing", function () {
    render(
      <MemoryRouter>
        <SignupForm></SignupForm>
      </MemoryRouter>
    );
    expect(screen.getByText("Username")).toBeInTheDocument();
  });

  it("matches snapshot", function () {
    const { asFragment } = render(
      <MemoryRouter>
        <SignupForm />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("displays 'Required' errors when submitted with no input values", async function () {
    render(
      <MemoryRouter>
        <SignupForm></SignupForm>
      </MemoryRouter>
    );
    const buttonEl = screen.getByRole("button");
    fireEvent.click(buttonEl);
    const requiredErrors = await screen.findAllByText("Required");
    expect(requiredErrors.length).toEqual(6);
  });

  it("allows user to sign up", async function () {
    const signup = jest.fn().mockResolvedValue({});

    render(
      <MemoryRouter>
        <SignupForm signup={signup} />
      </MemoryRouter>
    );

    act(() => {
      userEvent.type(screen.getByTestId("username"), "tester1");
      userEvent.type(screen.getByTestId("email"), "tester1@test.com");
      userEvent.type(screen.getByTestId("password"), "password");
      userEvent.type(screen.getByTestId("firstname"), "Tester");
      userEvent.type(screen.getByTestId("lastname"), "One");
      userEvent.type(screen.getByTestId("zipcode"), "36830");
      userEvent.click(screen.getByText("Create Account"));
    });

    await waitFor(() =>
      expect(signup).toBeCalledWith({
        username: "tester1",
        email: "tester1@test.com",
        password: "password",
        firstName: "Tester",
        lastName: "One",
        zipCode: "36830",
      })
    );
  });
});
