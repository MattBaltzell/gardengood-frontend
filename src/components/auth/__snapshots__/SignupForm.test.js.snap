// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupForm component tests matches snapshot 1`] = `
<DocumentFragment>
  <main
    class="Form"
  >
    <div
      class="Home__bg"
    />
    <h1>
      Sign Up
    </h1>
    <form
      action="#"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        autocomplete="username"
        class="text-input"
        data-testid="username"
        name="username"
        type="text"
        value=""
      />
      <label
        for="email"
      >
        Email
      </label>
      <input
        autocomplete="email"
        class="text-input"
        data-testid="email"
        name="email"
        type="text"
        value=""
      />
      <label
        for="password"
      >
        Password
      </label>
      <input
        class="password-input"
        data-testid="password"
        name="password"
        type="password"
        value=""
      />
      <label
        for="firstName"
      >
        First Name
      </label>
      <input
        class="text-input"
        data-testid="firstname"
        name="firstName"
        type="text"
        value=""
      />
      <label
        for="lastName"
      >
        Last Name
      </label>
      <input
        class="text-input"
        data-testid="lastname"
        name="lastName"
        type="text"
        value=""
      />
      <label
        for="zipCode"
      >
        Zip Code
      </label>
      <input
        class="text-input"
        data-testid="zipcode"
        name="zipCode"
        pattern="[0-9]{5}"
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Create Account
      </button>
    </form>
    <p>
      Already have an account? 
      <a
        href="/login"
      >
        Log in
      </a>
    </p>
  </main>
</DocumentFragment>
`;
